### PR TITLE
table.children and table.parents can now return table objects, not just names

### PR DESCRIPTION
**Pull request by dimitri-yatsenko**
Fix issue #821

### CODE REVIEW

File: datajoint/table.py [hunk 1]
```diff
@@ -116,29 +116,55 @@ def get_select_fields(self, select_fields=None):
...
-    def descendants(self):
-        return self.connection.dependencies.descendants(self.full_table_name)
-
-    def ancestors(self):
-        return self.connection.dependencies.ancestors(self.full_table_name)
+        :param as_objects: if False (default), the output is a dict describing the foreign keys. If True, return table objects.
+        :return: dict of tables with foreign keys referencing self or list of table objects if as_objects=True
+        """
+        nodes = dict((next(iter(self.connection.dependencies.children(k).items())) if k.isdigit() else (k, v))
+                     for k, v in self.connection.dependencies.children(self.full_table_name, primary).items())
```
Comment (guzman-raphael) [inline datajoint/table.py:141]: Could you elaborate why it is necessary to check if key is a string representation of a number? Unclear what is the overall issue this PR is trying to solve (aside from allowing objects to be returned and updates to dependency loading). Would you mind updating the `CHANGELOG.md` + and PR description?

File: datajoint/table.py [hunk 2]
```diff
@@ -116,29 +116,55 @@ def get_select_fields(self, select_fields=None):
...
+        if as_objects:
+            nodes = [FreeTable(self.connection, c) for c in nodes]
+        return nodes
+
+    def descendants(self, as_objects=False):
+        nodes = [node for node in self.connection.dependencies.descendants(self.full_table_name) if not node.isdigit()]
```
Comment (guzman-raphael) [inline datajoint/table.py:147]: Also unclear why this is a subset now than prior implementation.

**Review by guzman-raphael — CHANGES_REQUESTED** [review]
@dimitri-yatsenko Looks good but unclear on some points. Looking for a bit more context/clarification to complete the review.

File: datajoint/table.py (hunk 1 above)
Comment (dimitri-yatsenko) [reply datajoint/table.py:141]: When DataJoint creates the dependency graph, it introduces extra nodes when the foreign key attributes have different names than the primary key attributes that they reference.  This allows creating multi-graphs where any two nodes can be connected by multiple edges. The names of such nodes are composed of incremental numbers. Thus nodes composed of numbers represent "alias nodes" that do not represent a table.

File: datajoint/table.py (hunk 2 above)
Comment (dimitri-yatsenko) [reply datajoint/table.py:147]: prior implementation included the the alias nodes that represent renamed foreign keys. Now it only returns the nodes that represent tables.

File: datajoint/table.py (hunk 1 above)
Comment (dimitri-yatsenko) [reply datajoint/table.py:141]: updated CHANGELOG and version to 0.12.8

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,11 @@
 ## Release notes
 
+### 0.12.8 -- Dec 14, 2020
```
Comment (guzman-raphael) [inline CHANGELOG.md:3]:
```suggestion
### 0.12.8 -- Dec 21, 2020
```

**Review by guzman-raphael — CHANGES_REQUESTED** [review]
@dimitri-yatsenko Great work! Would you limit the maximum line character length to 95 (at least on any modified/introduced code)? PEP8 states 79 but really 95 is all that is required for a cleaner PR review on small screens. Looking to address this in codebase as it comes up.

Other than that, making a request to improve performance of `unite_master_parts` algorithm on case of high number of parts, doc updates, and style updates.

File: datajoint/dependencies.py [hunk 4]
```diff
@@ -1,9 +1,31 @@
...
 from .errors import DataJointError
 
 
+def unite_master_parts(lst):
+    """
+    reorder a list of tables names so that part tables immediately follow their master tables without breaking
+    the topological order.
+    Without this correction, a simple topological sort may insert other descendants between master and parts
+    :example:
+    _unite(['`s`.`a`', '`s`.`a__q`', '`s`.`b`', '`s`.`c`', '`s`.`c__q`', '`s`.`b__q`', '`s`.`d`', '`s`.`a__r`'])
```
[the remark this answers is not in the document]
Comment (guzman-raphael) [reply datajoint/dependencies.py]: Outdated now.

File: datajoint/dependencies.py [hunk 5]
```diff
@@ -1,9 +1,37 @@
...
+    for i in range(2, len(lst)):
+        name = lst[i]
+        match = re.match(r'(?P<master>`\w+`.`\w+)__\w+`', name)
+        if match:  # name is a part table
+            master = match.group('master')
+            for j in range(i-1, -1, -1):
+                if lst[j] == master + '`' or lst[j].startswith(master + '__'):
+                    # move from the ith position to the (j+1)th position
+                    del lst[i]
+                    lst = lst[:j+1] + [name] + lst[j+1:]
```
[the remark this answers is not in the document]
Comment (dimitri-yatsenko) [reply datajoint/dependencies.py:28]: good point!

**Comment by guzman-raphael** [comment]
@dimitri-yatsenko Sorry, missed this before... Would you update your PR description to reference #821 instead of #421; think this was a typo? Also, can you add reference to #179 since it is related?

File: datajoint/table.py [hunk 6]
```diff
@@ -116,29 +116,55 @@ def get_select_fields(self, select_fields=None):
...
-    def ancestors(self):
-        return self.connection.dependencies.ancestors(self.full_table_name)
+        :param as_objects: if False (default), the output is a dict describing the foreign keys. If True, return table objects.
+        :return: dict of tables with foreign keys referencing self or list of table objects if as_objects=True
+        """
+        nodes = dict((next(iter(self.connection.dependencies.children(k).items())) if k.isdigit() else (k, v))
+                     for k, v in self.connection.dependencies.children(self.full_table_name, primary).items())
+        if as_objects:
+            nodes = [FreeTable(self.connection, c) for c in nodes]
+        return nodes
```
[the remark this answers is not in the document]
Comment (dimitri-yatsenko) [reply datajoint/table.py:144]: This not quite equivalent. In the original code with `as_objects=False`, the function returns a dict.

**Review by guzman-raphael — CHANGES_REQUESTED** [review]
Final review with minor nits ahead of merge/release. Would you also ensure a max line length of 95? Noticeable in changes made in `datajoint/dependencies.py` and perhaps in other updates as well.

**Comment by guzman-raphael** [comment]
@dimitri-yatsenko Based on the tests, looks like you will need to upgrade the image for `fakeservices.datajoint.io` as well. I believe the latest is: `raphaelguzman/nginx:v0.0.13`.

File: datajoint/table.py [hunk 7]
```diff
@@ -11,7 +11,7 @@
 from .declare import declare, alter
 from .expression import QueryExpression
 from . import blob
-from .utils import user_choice
+from .utils import user_choice, OrderedDict
```
[the remark this answers is not in the document]
Comment (dimitri-yatsenko) [reply datajoint/table.py:14]: you are right. Fixed.

**Review by guzman-raphael — APPROVED** [review]

**Pull request merged**